File: PadmeCDR/code/transfer_files.py
#!/usr/bin/python

import os
import sys
import time
import getopt
import subprocess

def main(argv):

    try:
        opts,args = getopt.getopt(argv,"y:h")
    except getopt.GetoptError:
        print 'transfer_files [-y year] [-h]'
        sys.exit(2)

    YEAR = time.strftime("%Y",time.localtime())
    for opt,arg in opts:
        if opt == '-h':
            print 'transfer_files [-y year] [-h]'
            sys.exit()
        elif opt == '-y':
            YEAR = arg

    DATA_DIR = "rawdata/%s"%YEAR

    DAQ_SRV = "padmesrv1"
    DAQ_USER = "daq"
    DAQ_PATH = "/home/daq/DAQ"

    DAQ_ADLER32_CMD = "/home/daq/DAQ/tools/adler32"

    DAQ_SFTP = "sftp://%s%s"%(DAQ_SRV,DAQ_PATH)
    LNF_SRM = "srm://atlasse.lnf.infn.it:8446/srm/managerv2?SFN=/dpm/lnf.infn.it/home/vo.padme.org/testbeam"
    CNAF_SRM = "srm://storm-fe-archive.cr.cnaf.infn.it:8444/srm/managerv2?SFN=/padmeTape/testbeam"

    print "Getting list of raw data files for year %s on %s"%(YEAR,DAQ_SRV)
    daq_list = []
    for line in run_command("ssh -l %s %s ls %s/%s"%(DAQ_USER,DAQ_SRV,DAQ_PATH,DATA_DIR)):
        daq_list.append(line.rstrip())

    print "Getting list of raw data files for year %s on LNF disks"%YEAR
    lnf_list = []
    for line in run_command("gfal-ls %s/%s"%(LNF_SRM,DATA_DIR)):
        lnf_list.append(line.rstrip())

    print "Getting list of raw data files for year %s on CNAF tape library"%YEAR
    cnaf_list = []
    for line in run_command("gfal-ls %s/%s"%(CNAF_SRM,DATA_DIR)):
        cnaf_list.append(line.rstrip())

    # Make a full list and show which file is located where
    full_list = []
    full_list.extend(daq_list)
    full_list.extend(lnf_list)
    full_list.extend(cnaf_list)
    full_list = sorted(set(full_list))
    for rawfile in (full_list):
        print "%-30s %s%s%s"%(rawfile,
                              ("D" if rawfile in daq_list else "-"),
                              ("L" if rawfile in lnf_list else "-"),
                              ("C" if rawfile in cnaf_list else "-"))

    for rawfile in (full_list):

        #print "=== Checking file %s ==="%rawfile

        if ( rawfile in daq_list and not rawfile in lnf_list ):

            print "--- Getting ADLER32 checksum of %s ---"%rawfile
            cmd = "ssh -l %s %s %s %s/%s/%s"%(DAQ_USER,DAQ_SRV,DAQ_ADLER32_CMD,DAQ_PATH,DATA_DIR,rawfile)
            print("> %s"%cmd)
            for line in run_command(cmd):
                (a32_daq,fdummy) = line.rstrip().split()
            print "Source file ADLER32 CRC: %s"%a32_daq

            print "--- Starting copy of %s from DAQ to LNF ---"%rawfile
            #cmd = "gfal-copy -t 3600 -T 3600 --checksum ADLER32:%s --checksum-mode target -D\"SFTP PLUGIN:USER=%s\" -D\"SFTP PLUGIN:PRIVKEY=/home/%s/.ssh/id_rsa\" %s/%s/%s %s/%s/%s"%(a32_daq,DAQ_USER,os.environ['USER'],DAQ_SFTP,DATA_DIR,rawfile,LNF_SRM,DATA_DIR,rawfile);
            #cmd = "gfal-copy -t 3600 -T 3600 --checksum ADLER32:%s -D\"SFTP PLUGIN:USER=%s\" -D\"SFTP PLUGIN:PRIVKEY=/home/%s/.ssh/id_rsa\" %s/%s/%s %s/%s/%s"%(a32_daq,DAQ_USER,os.environ['USER'],DAQ_SFTP,DATA_DIR,rawfile,LNF_SRM,DATA_DIR,rawfile);
            cmd = "gfal-copy -t 3600 -T 3600 -D\"SFTP PLUGIN:USER=%s\" -D\"SFTP PLUGIN:PRIVKEY=/home/%s/.ssh/id_rsa\" %s/%s/%s %s/%s/%s"%(DAQ_USER,os.environ['USER'],DAQ_SFTP,DATA_DIR,rawfile,LNF_SRM,DATA_DIR,rawfile);
            print("> %s"%cmd)
            for line in run_command(cmd):
                print(line.rstrip())
            # Need to learn how to handle gfal_copy errors, timeouts, etc...

            print("Copy to LNF was successful: adding %s to LNF list"%rawfile)
            lnf_list.append(rawfile)

        if ( rawfile in lnf_list and not rawfile in cnaf_list ):

            print "--- Starting copy of %s from LNF to CNAF ---"%rawfile
            cmd = "gfal-copy -t 3600 -T 3600 --checksum ADLER32 %s/%s/%s %s/%s/%s"%(LNF_SRM,DATA_DIR,rawfile,CNAF_SRM,DATA_DIR,rawfile);
            print("> %s"%cmd)
            for line in run_command(cmd):
                print(line.rstrip())
            # Need to learn how to handle gfal_copy errors, timeouts, etc...

            print("Copy to CNAF was successful: adding %s to CNAF list"%rawfile)
            cnaf_list.append(rawfile)

def run_command(command):
    p = subprocess.Popen(command,stdout=subprocess.PIPE,stderr=subprocess.STDOUT,shell=True)
    return iter(p.stdout.readline, b'')

# Execution starts here
if __name__ == "__main__":
   main(sys.argv[1:])

File: PadmeCDR/code/transfer_files.pl
#!/usr/bin/perl

use Getopt::Std;

printf "Running on node $ENV{'HOSTNAME'} as user $ENV{'USER'} in dir $ENV{'HOME'}\n";

$YEAR = 1900+(localtime)[5];
getopts('y:');
if ($opt_y) {
    $YEAR = $opt_y;
}
printf "Will transfer files for year $YEAR\n";

$DATA_DIR = "rawdata/$YEAR";

$DAQ_SRV = "padmesrv1";
$DAQ_USER = "daq";
$DAQ_PATH = "/home/daq/DAQ";

$DAQ_ADLER32_CMD = "/home/daq/DAQ/tools/adler32";

$DAQ_SFTP = "sftp://$DAQ_SRV$DAQ_PATH";
$LNF_SRM = "srm://atlasse.lnf.infn.it:8446/srm/managerv2?SFN=/dpm/lnf.infn.it/home/vo.padme.org/testbeam";
$CNAF_SRM = "srm://storm-fe-archive.cr.cnaf.infn.it:8444/srm/managerv2?SFN=/padmeTape/testbeam";

print "Getting list of raw data files for year $YEAR on $DAQ_SRV\n";
my @daq_list = `ssh -l $DAQ_USER $DAQ_SRV ls $DAQ_PATH/$DATA_DIR`;
chomp(@daq_list);
my %daq_list = map { $_ => 1 } @daq_list;

print "Getting list of raw data files for year $YEAR on LNF disks\n";
my @lnf_list = `gfal-ls $LNF_SRM/$DATA_DIR`;
chomp(@lnf_list);
my %lnf_list = map { $_ => 1 } @lnf_list;

print "Getting list of raw data files for year $YEAR on CNAF tape library\n";
my @cnaf_list = `gfal-ls $CNAF_SRM/$DATA_DIR`;
chomp(@cnaf_list);
my %cnaf_list = map { $_ => 1 } @cnaf_list;

# Make a full list and show which file is located where
@full_list = ();
push(@full_list,@daq_list);
push(@full_list,@lnf_list);
push(@full_list,@cnaf_list);
@full_list = sort @full_list;
$i = 0;
$uniq_list[0] = $full_list[0];
foreach my $file (@full_list) {
    if ($file ne $uniq_list[$i]) {
	push(@uniq_list,$file);
	$i++;
    }
}
foreach my $file (@uniq_list) {
    printf("%-30s ",$file);
    if ( exists($daq_list{$file})  ) { printf "D"; } else { printf "-"; }
    if ( exists($lnf_list{$file})  ) { printf "L"; } else { printf "-"; }
    if ( exists($cnaf_list{$file}) ) { printf "C"; } else { printf "-"; }
    printf "\n";
}

#foreach my $file ("run_628.root") {
foreach my $file (@uniq_list) {

    printf "\n=== Checking file $file ===\n";

    # If file exists on DAQ and not at LNF, copy it
    if ( exists($daq_list{$file}) and not exists($lnf_list{$file}) ) {

	printf "Starting copy of $file from DAQ to LNF\n";

	# Copy from PADME on-line server to LNF storage system
	$cmd = "gfal-copy -t 3600 -T 3600 -D\"SFTP PLUGIN:USER=$DAQ_USER\" -D\"SFTP PLUGIN:PRIVKEY=/home/$ENV{'USER'}/.ssh/id_rsa\" $DAQ_SFTP/$DATA_DIR/$file $LNF_SRM/$DATA_DIR/$file";
	printf "> $cmd\n";
	#system(split(/ /,$cmd));
	system($cmd);

	# Verify ADLER32 checksum

	$a32_daq_out = `ssh -l $DAQ_USER $DAQ_SRV $DAQ_ADLER32_CMD $DAQ_PATH/$DATA_DIR/$file`;
	chomp $a32_daq_out;
	$a32_daq = (split(/ /,$a32_daq_out))[0];

	$a32_lnf_out = `gfal-sum $LNF_SRM/$DATA_DIR/$file ADLER32`;
	chomp $a32_lnf_out;
	$a32_lnf = (split(/ /,$a32_lnf_out))[1];

	if ( $a32_daq ne $a32_lnf ) {
	    printf "ADLER32 CRC ***ERROR***: DAQ $a32_daq LNF $a32_lnf\n";
	} else {
	    printf "ADLER32 CRC OK: DAQ $a32_daq LNF $a32_lnf\n";
	    # Copy was successful: add file to list at LNF
	    $lnf_list{$file} = 1;
	    # Add DAQ->LNF post-transfer processing here
	}

    }

    # If file exist at LNF but not at CNAF, copy it
    if ( exists($lnf_list{$file}) and not exists($cnaf_list{$file}) ) {

	printf "Starting copy of $file from LNF to CNAF\n";

	# Copy from LNF storage system to CNAF tape library using gfal-copy
	$cmd = "gfal-copy -t 3600 -T 3600 $LNF_SRM/$DATA_DIR/$file $CNAF_SRM/$DATA_DIR/$file";
	printf "> $cmd\n";
	system(split(/ /,$cmd));

	# Verify ADLER32 checksum

	$a32_lnf_out = `gfal-sum $LNF_SRM/$DATA_DIR/$file ADLER32`;
	chomp $a32_lnf_out;
	$a32_lnf = (split(/ /,$a32_lnf_out))[1];

	$a32_cnaf_out = `gfal-sum $CNAF_SRM/$DATA_DIR/$file ADLER32`;
	chomp $a32_cnaf_out;
	$a32_cnaf = (split(/ /,$a32_cnaf_out))[1];

	if ( $a32_lnf ne $a32_cnaf ) {
	    printf "ADLER32 CRC ***ERROR***: LNF $a32_lnf CNAF $a32_cnaf\n";
	} else {
	    printf "ADLER32 CRC OK: LNF $a32_lnf CNAF $a32_cnaf\n";
	    # Copy was successful: add file to list at LNF
	    $cnaf_list{$file} = 1;
	    # Add LNF->CNAF post-transfer processing here
	}

    }

}
